text/2020-01-30-key-spaces.md

```diff
@@ -0,0 +1,181 @@
+
+The client can register for additional key spaces.
+
+#### Metadata operations
+
+POST /keyspaces
+
+requires:
+
+* application name
+
+optional:
+
+* key space name
+* description
+* prefix
+
+This API will return the key space object, which includes an ID. Key spaces have an ID to allow their name to be changed.
+Clients are recommended to provide a key space name, but the API can also auto-generate one.
+
+The usage of an application name as an identity may end up being a de-normalization of data in a future iteration when identity is added for authentication and there are identity ids.
+I suggest that we can add an identity id as an alternative field to an application name. Similarly, the output key space object can include the identity id.
+
+Addtional GET APIs will also return keyspace objects
+
+* Lookup by ID, key space name
+* List all key spaces
+
+POST /keyspaces/{id}
+
+requires id and description. Only the description field can be updated.
+
+DELETE /keyspaces/{id}
+
+requires id. A key space may only be deleted after all its used keys are deleted.
+Before the delete operation is completed, it requires PD to first notify all TiKV that would receive requests for the key space of the deletion.
+
+#### TiKV storage operations
+
+Requests to TiKV will have a new field "key space", which is filled with the key space id.
+TiKV looks up the key space id in PD to translate it to a key prefix.
+TiKV will then prepend the key prefix to all key usage in the request.
+It will strip the prefix from returned keys.
+Note that this lookup with PD will be cached.
+The cache can only expire if the key space is deleted because the key space id and prefix do not change.
+
+#### Key Prefix specification or compression
+
+Normally, a client registers for a keyspace with a name, but the actual key prefix is transparently assigned, known only to TiKV.
+This allows TiKV to change the prefix when doing a restore and also allows it to perform key prefix compression.
+Note that some storage engines may already be able to perform prefix compression.
+For backwards compatibility purposes, a client is also allowed to be registered with an explicit key prefix.
+
+#### The default key space
+
+A TiKV client may ignore key spaces. In a fresh TiKV deployment with this feature in place, that client will be using the default key space.
+A configuration option `disable-default-key-space` when set to `true` forces clients to specify a key space rather than using the default key space. 
+
+#### Backwards compatibility
+
+For an existing deployment we will need an upgrade process.
+We can have a configuration flag "allow-unmanaged-key-space".
+For a fresh deployment this will be `false`, but if there is an existing deployment it should be `true`.
+When true, clients that do not specify a key space will continue using the key space the same way they do now.
+When false, clients that do not specify a key space will use the default key space.
```

Comment:
What is the prefix key for default key space?

Reply:
How about "default"?